test/unit/lua_msgpack.c

```diff
@@ -0,0 +1,117 @@
+#include <lualib.h>
+
+#include "core/assoc.h"
+#include "core/cord_buf.h"
+#include "core/fiber.h"
+#include "core/memory.h"
+
+#include "lua/msgpack.h"
+#include "lua/serializer.h"
+
+#include "mpstream/mpstream.h"
+
+#include "small/ibuf.h"
+
+#define UNIT_TAP_COMPATIBLE 1
+#include "unit.h"
+
+static void
+mpstream_error_mock(void *ctx)
+{
+	(void)ctx;
+}
+
+/**
+ * Checks that translation of first-level `MP_MAP` keys is done correctly.
+ */
+static void
+test_translation_in_encoding(lua_State *L)
+{
+	plan(3);
+	header();
+
+	struct mh_strnu32_t *translation = mh_strnu32_new();
+	const char alias[] = "x";
+	struct mh_strnu32_node_t node = {
+		.str = alias,
+		.len = lengthof(alias) - 1,
+		.hash = mh_strn_hash(alias, lengthof(alias) - 1),
+		.val = 0
+	};
+	mh_strnu32_put(translation, &node, NULL, NULL);
+
+	struct ibuf *ibuf = cord_ibuf_take();
+	struct mpstream stream;
+	mpstream_init(&stream, ibuf, ibuf_reserve_cb, ibuf_alloc_cb,
+		      mpstream_error_mock, L);
+
+	lua_createtable(L, 0, 1);
+	lua_pushboolean(L, true);
+	lua_setfield(L, 1, alias);
+	luamp_encode_with_translation(L, luaL_msgpack_default, &stream, 1,
+				      translation);
+	lua_pop(L, 1);
+	mpstream_flush(&stream);
+	ok(strncmp(ibuf->buf, "\x81\x00\xc3", ibuf_used(ibuf)) == 0,
+	   "first-level MP_MAP key is translated");
+	ibuf_reset(ibuf);
+	mpstream_reset(&stream);
+
+	lua_createtable(L, 0, 1);
+	lua_createtable(L, 0, 1);
+	lua_pushboolean(L, true);
+	lua_setfield(L, -2, alias);
+	lua_setfield(L, -2, "k");
+	luamp_encode_with_translation(L, luaL_msgpack_default, &stream, 1,
+				      translation);
+	lua_pop(L, 1);
+	mpstream_flush(&stream);
+	ok(strncmp(ibuf->buf, "\x81\xa1k\x81\xa1x\xc3",
+		   ibuf_used(ibuf)) == 0,
+	   "only first-level MP_MAP key is translated");
+	ibuf_reset(ibuf);
+	mpstream_reset(&stream);
+
+	lua_createtable(L, 0, 1);
+	lua_pushnumber(L, 0);
+	lua_pushboolean(L, true);
+	lua_settable(L, -3);
+	luamp_encode_with_translation(L, luaL_msgpack_default, &stream, 1,
+				      translation);
```

Comment:
What happens if the Lua table contains both the alias and its value `t = {x = 10, [0] = 20}`?

Reply:
Discussed with @locker and @alyapunov: decided to declare this unspecified behavior.